### PR TITLE
KDL based Movement in Cartesian Space

### PR DESCRIPTION
**Pull request by BoWangFromMars**
KDL(Kinematics and Dynamics Library) installation process is introduced in this branch. Based on forward kinematics and inverse kinematics provided by KDL, a demo, which makes the robot move in z axis in Cartesian space, is added in this branch. Please see the detailed information in pdf file "KDL installation and demo.pdf" in the directory './lbr_demos_fri_ros2_cpp/doc'.

### CODE REVIEW

**Comment by mhubii** [comment]
thank you for the PR @BoWangFromMars. I would create a development branch so you can re-target the PR against this branch. Then I can make some modifications to the PR on the dev branch.

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            return;
+        }
+
+	public:
+		JointPositionPublisher():Node("joint_position_publisher_node"), amplitude_(0.05), frequency_(0.5), sampling_time_(0.01)
+		{
+            phase_ = 0.0; 
+            count_= 0.0;
+
+            std::string urdf_file_path = "/home/nearlab-iiwa/lbr-stack/src/lbr_fri_ros2_stack/lbr_description/urdf/iiwa7/my_iiwa7.urdf"; // path of your robot urdf file
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp]: would it be possible to load this from the `/robot_description` parameter instead

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+class JointPositionPublisher:public rclcpp::Node
+{
+	private:
+		rclcpp::Publisher<lbr_fri_msgs::msg::LBRPositionCommand>::SharedPtr publisher_;
+        rclcpp::Subscription<lbr_fri_msgs::msg::LBRState>::SharedPtr subscriber_;
+
+        KDL::Chain chain_; // robot kinematics chain exetracted from robot URDF file
+
+        double amplitude_; // rad
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp:30]: don't think these variables are used anywhere

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+using namespace std::chrono_literals; // this namespace to enable we can use 500ms to represent 500 milliseconds in our code.
+using std::placeholders::_1;
+
+lbr_fri_msgs::msg::LBRState current_robot_state; // store current robot state, including measured joint positions
+KDL::Frame initial_cartesian_position; // we want to move the end effector in sine curve in z-direction in Cartesian space, this is the initial position 
+
+class JointPositionPublisher:public rclcpp::Node
+{
+	private:
+		rclcpp::Publisher<lbr_fri_msgs::msg::LBRPositionCommand>::SharedPtr publisher_;
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp]: make sure formatting is correct and line limit is 100 characters

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+#include "kdl/chainiksolverpos_lma.hpp" // for inverse kinematics
+#include "lbr_fri_msgs/msg/lbr_position_command.hpp"
+#include "lbr_fri_msgs/msg/lbr_state.hpp"
+#include "friClientIf.h"
+
+
+using namespace std::chrono_literals; // this namespace to enable we can use 500ms to represent 500 milliseconds in our code.
+using std::placeholders::_1;
+
+lbr_fri_msgs::msg::LBRState current_robot_state; // store current robot state, including measured joint positions
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp]: can these be stored as members

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+	private:
+        /**
+         * @function: callback function for Robot State Subscriber
+        */
+        void topic_callback(const lbr_fri_msgs::msg::LBRState& msg)
+        {
+            current_robot_state = msg;
+            
+            if(count_ < 200) // in first 2 seconds, to make 'initial_position_command' be 'measured_joint_position'
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp]: might be better to introduce `is_init` bool or turn `current_robot_state` into unique pointer and check for null

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+         * @param current_joint_positions current joint positions
+         * @return joint positions command 
+        */
+        lbr_fri_msgs::msg::LBRPositionCommand computeInverseKinematics(KDL::Chain& chain, KDL::Frame& desired_end_effector_pose, KDL::JntArray& current_joint_positions)
+        {
+            KDL::ChainIkSolverPos_LMA ik_solver(chain);
+            KDL::JntArray result_joint_positions = KDL::JntArray(chain.getNrOfJoints());
+            lbr_fri_msgs::msg::LBRPositionCommand joint_position_command;
+
+            auto start = std::chrono::high_resolution_clock::now();
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp]: maybe only measure time on debug

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+                }
+                initial_cartesian_position = computeForwardKinematics(chain_, joint_position);
+            }
+            else
+            {
+                if (current_robot_state.session_state == KUKA::FRI::COMMANDING_ACTIVE) 
+                {
+                    KDL::Frame cartesian_position_command = initial_cartesian_position;
+
+                    cartesian_position_command.p = KDL::Vector(initial_cartesian_position.p.x(), initial_cartesian_position.p.y(), initial_cartesian_position.p.z() + amplitude_ * sin(phase_));
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_path_planning_node.cpp]: oh okay I see, this performs oscillation in Cartesian space, could you maybe add a Cartesian pose topic or even better a Twist topic (to indicate velocity commands) instead? Twist commands might need some code changes

**Review by mhubii — COMMENTED** [review]
Reviewed your PR. Could you please amend changes. Not all suggestions might make sense, so feel free to keep things that make sense to you

**Comment by mhubii** [comment]
so ideally the robot descriptions will be loaded from the `/robot_description` topic, and there is no need for the additional `.urdf` files. Hope that makes sense. Alternatively, the robot description could be parsed as a parameter to the node.

### From Topic
Create subscription to `/robot_description`. `robot_state_publisher` publishes the robot description with QoS `transient local` topic, see e.g. https://docs.ros.org/en/rolling/Concepts/Intermediate/About-Quality-of-Service-Settings.html:

- Durability
  - Transient local: the publisher becomes responsible for persisting samples for “late-joining” subscriptions.

### From Parameter
Refer e.g. https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/036b23b2f91af70e1ca490e0c81660b6b17ae114/lbr_demos/lbr_demos_fri_ros2_advanced_cpp/src/admittance_control_node.cpp#L17

File: lbr_description/urdf/iiwa14/iiwa14.urdf [hunk 8]
```diff
@@ -0,0 +1,229 @@
+<?xml version="1.0" ?>
```
Comment (mhubii) [inline lbr_description/urdf/iiwa14/iiwa14.urdf]: these files are redundant

File: lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_pose_node.cpp [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+      if(robot_name != "iiwa7" && robot_name != "iiwa14" && 
+         robot_name != "med7" && robot_name != "med14")
+      {
+        std::cerr << "Wrong robot name. " << std::endl;
+      }
+
+      // get the path of urdf file 
+      std::filesystem::path current_working_directory = std::filesystem::current_path();
+      std::string urdf_file_path = current_working_directory.string()
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_pose_node.cpp]: getting the current path might not work in general

Comment (BoWangFromMars) [reply lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_pose_node.cpp]: Hello, @mhubii,  thank you very much for your suggestion and QoS documentation, I learn a lot from it. I modify my code following your advice today and then upload it again after I test it.

Comment (mhubii) [reply lbr_demos/lbr_demos_fri_ros2_cpp/src/cartesian_pose_node.cpp]: awesome, happy to merge once the `/robot_description` is loaded from topic / parameter and `.urdf` files are removed

File: lbr_demos/lbr_demos_fri_ros2_cpp/doc/KDL installation.pdf
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/doc/KDL installation.pdf:1]: one more question regarding this doc. KDL is usually installed through `rosdep` as listed in `package.xml` files, refer e.g. https://github.com/lbr-stack/lbr_fri_ros2_stack/blob/036b23b2f91af70e1ca490e0c81660b6b17ae114/lbr_demos/lbr_demos_fri_ros2_advanced_cpp/package.xml#L18

File: lbr_demos/lbr_demos_fri_ros2_cpp/package.xml [hunk 10]
```diff
@@ -12,6 +12,12 @@
   <depend>fri_vendor</depend>
   <depend>lbr_fri_msgs</depend>
   <depend>rclcpp</depend>
+  <depend>tf2_ros</depend>
+  <depend>urdf</depend>
+  <depend>geometry_msgs</depend>
+  <depend>sensor_msgs</depend>
+  <depend>orocos_kdl</depend>
```
Comment (mhubii) [inline lbr_demos/lbr_demos_fri_ros2_cpp/package.xml]: this dependency should be replaced by `orocos_kdl_vendor`

**Review by mhubii — COMMENTED** [review]
some more dependency hints

File: lbr_demos/lbr_demos_fri_ros2_cpp/package.xml (hunk 10 above)
Comment (BoWangFromMars) [reply lbr_demos/lbr_demos_fri_ros2_cpp/package.xml]: okay, I get it, thank you.

**Comment by mhubii** [comment]
thank you again for the PR @BoWangFromMars. I merged this into the `dev` branch and will update in #170.

If you want to make amend changes yourself, please add a PR to the dev branch again.

**Pull request merged**